### DISCUSSION TITLE
Dynamic rules / (unknown destination account) from ABN Amro

### DISCUSSION THREAD

**Discussion by FischerLouis** [firefly-iii/firefly-iii#7409] [Firefly III] [answered]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)
- [X] My request is not listed as [a very good idea, but unfortunately...](https://docs.firefly-iii.org/firefly-iii/more-information/what-its-not/)
- [X] I've used [the search](https://github.com/firefly-iii/firefly-iii/issues?q=is%3Aissue) and this has not been requested before.

### Description

- Currently importing my data via Nordigen from the Dutch bank ABN Amro.
- A good chunk of my transactions is done via debit card (widely used in the country), unfortunately categorised as (unknown destination account) in the import...
- Checking the description field of such transaction, I can see that all information is actually there to build a proper categorisation. Example: "BEA, Betaalpas Bakker Van Vessem,PASXXX, 14.04.23/10:01 HAARLEM"
- I've been looking into the documentation, existing issues on Github and generally on Google, without finding a clear way on how to handle this and can't believe I'm the only one concerned by this (let me know if I missed anything here).

### Solution

- It seems like a simple string split and storing part of the output into a specified field would cover the issue. Perhaps already supported but I missed it?
- Does the rule engine support such dynamic mapping? Going through it I could only see static rules.

I understand the overall philosophy of keeping the engine simple, but unsure if such rules is considered too complex or not.

### What are alternatives?

- Perhaps this cleaning step can be done at another level (from Nordigen?). I would appreciate your guidance on the matter if any (otherwise I'm a bit stuck using the tool).
- It seems technically possible to leverage the API to batch edit all my transactions, but that sounded a bit overkill and wanted to make sure it was the only option.

### Additional context

Thanks for your input and for the tool! Really happy with it otherwise, would definitely invest more time into it if this bit was tackled.

**Comment by JC5** [comment]
I'm Dutch myself, I know how ABN works 😉. Having said that, this isn't something the rule engine will be able to fix.

**Comment by FischerLouis** [comment]
Thanks for your feedback!
Any advice on how such transactions could be properly categorised?
Is building a custom batch edit logic via the APIs the only option here or do you see another way to do this?
Thanks again.

**Comment by JC5** [comment]
I use webhooks myself to parse new transactions and push updates back.

**Comment by FischerLouis** [comment]
Ah good one, I missed that we had a webhook triggering on transaction creation. 
Sounds doable then, thanks!

**Comment by FischerLouis** [comment]
Quickly piggy-backing on this one. I'm able leverage the transaction creation webhook and the `PUT /v1/transactions/{id}` endpoint to adjust my transaction import (thanks again for the pointer).
For context: I'm simply parsing the `description` field, passing the relevant substring to `destination_name` field.

Although I'm getting a HTTP 200 on the PUT with adjusted `destination_name` field, the transaction is not updated at all (checking doing a GET right after).
Are there some fields protected on the transaction model? If so, is the only option to delete then re-create the transaction?

Thanks again for your help.

**Reply by JC5** [reply]
There's no special protection on any fields. Here's a curl request that works on the demo site?

```
curl --location --request PUT 'https://demo.firefly-iii.org/api/v1/transactions/309' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer ey.....' \
--data '{
    "transactions": [
        {
            "destination_name": "new name again!"
        }
    ]
}'
```

**Reply by JC5** [reply]
PS: if you also submit the (old) `destination_id` or `destination_iban` nothing will change, you have to drop those fields from your request.

**Reply by FischerLouis** [reply]
Thanks! That's helpful. I am indeed passing the `destination_name` field alone (not the other fields).
Trying out with the demo site works for me as well, quite odd then.

From my end I can update other fields via the PUT endpoint (ex: `description` field), so the overall logic seems to work. 
However passing `destination_name` field is somehow dropped.

Testing on the demo site, I can see that passing various `destination_name` values increments (or matches if existing) the `destination_id` field, is that correct?

The transactions I'm dealing with are imported from ABN via Nordigen, perhaps some fields / internal validation logics are preventing the update?
See below the raw transaction before update, let me know if anything comes to mind. I'm wondering if a destination_id should be provided for instance, to drop the link to the _(unknown destination account)_ value.

```
"transactions": [
        {
          "user": "1",
          "transaction_journal_id": "445",
          "type": "withdrawal",
          "date": "2023-04-17T00:00:00+02:00",
          "order": 0,
          "currency_id": "1",
          "currency_code": "EUR",
          "currency_name": "Euro",
          "currency_symbol": "\\u20ac",
          "currency_decimal_places": 2,
          "foreign_currency_id": null,
          "foreign_currency_code": null,
          "foreign_currency_symbol": null,
          "foreign_currency_decimal_places": 0,
          "amount": "17.200000000000",
          "foreign_amount": null,
          "description": "BEA, Betaalpas XXX",
          "source_id": "1",
          "source_name": "ABN - Shared account",
          "source_iban": "",
          "source_type": "Asset account",
          "destination_id": "7",
          "destination_name": "(unknown destination account)",
          "destination_iban": null,
          "destination_type": "Expense account",
          "budget_id": null,
          "budget_name": null,
          "category_id": null,
          "category_name": null,
          "bill_id": null,
          "bill_name": null,
          "reconciled": false,
          "notes": null,
          "tags": [
            "Data Import on 2023-04-19 @ 21:55"
          ],
          "internal_reference": null,
          "external_id": "0415081221128089",
          "original_source": "ff3-v6.0.5|api-v2.0.1",
          "recurrence_id": null,
          "recurrence_total": null,
          "recurrence_count": null,
          "bunq_payment_id": null,
          "external_url": null,
          "import_hash_v2": "3cb040b455...1eb1f1",
          "sepa_cc": null,
          "sepa_ct_op": null,
          "sepa_ct_id": null,
          "sepa_db": null,
          "sepa_country": null,
          "sepa_ep": null,
          "sepa_ci": null,
          "sepa_batch_id": null,
          "interest_date": null,
          "book_date": null,
          "process_date": null,
          "due_date": null,
          "payment_date": null,
          "invoice_date": null,
          "longitude": null,
          "latitude": null,
          "zoom_level": null,
          "has_attachments": false
        }
      ]
```

**Reply by JC5** [reply, marked as answer]
> Testing on the demo site, I can see that passing various destination_name values increments (or matches if existing) the destination_id field, is that correct?

Yes, it creates a new expense account if the name is new, otherwise it recycles an existing one.

> The transactions I'm dealing with are imported from ABN via Nordigen, perhaps some fields / internal validation logics are preventing the update?

No, but you can check out the logs of your container to see what's happening behind the screens. https://docs.firefly-iii.org/firefly-iii/faq/other/#how-do-i-enable-debug-mode

>  I'm wondering if a destination_id should be provided for instance, to drop the link to the (unknown destination account) value.

You should NOT provide the destination_id. "(unknown destination account)" is just another expense account and is technically the same as all the others, except that it's called "(unknown destination account)".

1. If you provide destination_id, it will overrule whatever is in destination_name.
2. If you provide destination_name only, Firefly III will search for (and create) the destination expense account.

**Reply by FischerLouis** [reply]
Thank you 🙏. Your guidance helped me identified the root cause of the issue (input payload was simply wrongly formatted, causing the update to be dropped).

Everything's working like a charm now 👍!
